Issue-Label-Bot-v2/notebooks/mlp.py

```diff
@@ -0,0 +1,171 @@
+        self.precision_thre = precision_thre
+        self.prob_thre = prob_thre
+        self.recall_thre = recall_thre
+        self.counter = counter
+        self.label_columns = label_columns
+        self.precision = None
+        self.recall = None
+        self.exclusion_list = None
+
+    def fit(self, X, y):
+        self.clf.fit(X, y)
+
+    def predict_proba(self, X):
+        return self.clf.predict_proba(X)
+
+    def calculate_auc(self, y_holdout, predictions):
+        if not self.counter or not self.label_columns:
+            raise Exception('Need to set counter and label columns to calculate AUC')
+        auc_scores = []
+        counts = []
+
+        for i, l in enumerate(self.label_columns):
+            y_hat = predictions[:, i]
+            y = y_holdout[:, i]
+            auc = roc_auc_score(y_true=y, y_score=y_hat)
+            auc_scores.append(auc)
+            counts.append(self.counter[l])
+
+        df = pd.DataFrame({'label': self.label_columns,
+                           'auc': auc_scores,
+                           'count': counts})
+        weightedavg_auc = df.apply(lambda x: x.auc * x['count'], axis=1).sum() / df['count'].sum()
+        print(f'Weighted Average AUC: {weightedavg_auc}')
+        return df, weightedavg_auc
+
+    def calculate_max_range_count(self, prob):
+        thresholds_lower = [0.1 * i for i in range(11)]
+        thresholds_upper = [0.1 * (i+1) for i in range(10)] + [1]
+        max_range_count = [0] * 11  # [0,0.1), [0.1,0.2), ... , [0.9,1), [1,1]
+        for i in prob:
+            max_range_count[int(max(i) // 0.1)] += 1
+
+        df = pd.DataFrame({'l': thresholds_lower,
+                           'u': thresholds_upper,
+                           'count': max_range_count})
+        return df, max_range_count
+
+    def calculate_argmax_result(self, y_true, y_pred, prob_thre=None):
+        if prob_thre:
+            self.prob_thre = prob_thre
+
+        y_pred_argmax = np.argmax(y_pred, axis=1)
+        y_pred_prob = np.max(y_pred, axis=1)
+
+        total_true = np.array([0] * len(y_pred[0]))
+        total_pred_true = np.array([0] * len(y_pred[0]))
+        pred_correct = np.array([0] * len(y_pred[0]))
+        for i in range(len(y_pred_argmax)):
+            y_true_label = np.where(y_true[i] == 1)[0]
+            total_true[y_true_label] += 1
+
+            if y_pred_prob[i] < self.prob_thre:
+                continue
+
+            pred_label = y_pred_argmax[i]
+            if pred_label in y_true_label:
+                pred_correct[pred_label] += 1
+            total_pred_true[pred_label] += 1
+
+        self.precision = pred_correct / total_pred_true
+        self.recall = pred_correct / total_true
+
+        df = pd.DataFrame({'label': self.label_columns,
+                           'precision': self.precision,
+                           'recall': self.recall})
+        return df, self.precision, self.recall
+
+    def find_best_prob_thre(self, y_true, y_pred):
```

Comment:
I'm really confused about this function.  The precision and recall threshold should be supplied by the user, as there isn't an objective way to automatically choose the best precision and recall thresholds automatically

Reply:
In this function, I would like to find a threshold which achieves more number of labels with precision larger than the predefined threshold, precision_thre.
If two thresholds have the same number of labels with precision larger than the precision threshold, I would choose the one with more labels having recall larger than the recall threshold.
That is because users may be more sensitive to false positive error and the recall is quite low when we set a higher threshold as I only predict one label at this moment.
This is not a good way to find the best threshold and we may need to figure out how to set a threshold to predict multiple labels instead of only one label.
If you are still confused, please let me know and we can have a phone call. Thanks.

Reply:
I will update to predict multiple labels with the predefined threshold, which much makes sense. So, the recall when `threshold=0.0` should be 100%.

Reply:
@abcdefgs0324 just to recap, this is the way I was thinking to manage the thresholds:

1.  Have the user supply their desired precision.  Default: 70%
2.  Have the user supply their minimum recall at this precision. Default 50%
3.  Find the threshold that meets this minimum standard.  There is a precision-recall tradeoff so the higher the precision the lower the recall these two metrics are strictly inversely related.   **This is why I'm confused why there is a search for the best threshold, as it is deterministic based on the precision / recall that the user chooses** . Keep in mind that the user will rarely choose this and the defaults will probably control 99% of the time.  

Let me know your thoughts or where I might be missing something.  Thanks!

Reply:
Also can you clarify what you mean when you say you are predicting only one label?  Can you shed some light on why?  I have a hunch that some of the confusion has to do with that

Reply:
I have updated the code to predict multiple labels. The procedure is that we set a threshold `prob_thre`, and all labels with probabilities larger than this threshold will be predicted `True`. You can see [the function `calculate_result`](https://github.com/abcdefgs0324/Issue-Label-Bot/blob/22dcf6ee2acf5810262f5745cee1e24a20c39199/Issue-Label-Bot-v2/notebooks/mlp.py#L91) in `mlp.py`.

Yes, the user will supply two values for precision and recall. And this function `find_best_prob_thre` is to find the threshold that meets precision and recall standards. However, there is a problem that we can not have a threshold to let all labels meet both of the precision and recall minimum.
You can see `Label_k8s_issues_with_MLP.ipynb` (cell 45), when `prob_thre=0.0`, some labels such as `sig/vmware` and `sig/service-catalog` have quite low precision. And it is impossible to let these labels to meet the precision minimum. So in my current implementation, I find a threshold to let more number of labels to meet both of precision and recall minimum and this is what this function does.

For example, assume when `prob_thre=0.5`, we have 10 labels that meet both of minimum. And if increasing the `prob_thre` only decreases the number of labels meeting two minimum, this function will choose `prob_thre=0.5` to be its best threshold.

Reply:
> So in my current implementation, I find a threshold to let more number of labels to meet both of precision and recall minimum and this is what this function does.

What do you mean by more number of labels?  more than what?  I was expecting that if labels do not meet the precision / threshold criteria you exclude them all together and do not offer any predictions on those labels, since the model would create more noise than value.  

> For example, assume when prob_thre=0.5, we have 10 labels that meet both of minimum.

I find it confusing that `prob_thre` is a parameter.  prob_thre should be something you calculate internally from the precision and recall thresholds, but **not** something that is specified by the user.  Perhaps this is an additional source of confusion?  

The api that expect is
```py
# set desired precision and recall
estimator.set_precision_recall(precision=<float>, recall=<float>)

# your estimator will return a list of labels.  If a label cannot be predicted with 
# precision/recall that means that your model is not good enough to opine on those 
# labels and should simply ignore those particular labels.
estimator.predict(body:str, title:str) -> List
```

Also, please consider using [sklearn's built in utility](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html) for finding the proper threshold if possible.  That might simplify the code and make your life easier!